pyanaconda/kickstart_dispatcher/element.py

```diff
@@ -0,0 +1,239 @@
+        """
+        self._elements.append(element)
+
+    @property
+    def all_elements(self):
+        """List of all elements in the container.
+
+        :return: list of all elements in the container
+        :rtype: list(KickstartElement)
+        """
+
+        return list(self._elements)
+
+    def get_elements(self, commands=None, sections=None, addons=None):
+        """Returns selected elements.
+
+        :param commands: names of commands to be returned
+        :type commands: list
+        :param sections: names of sections to be returned
+        :type sections: list
+        :param addons: names of addons to be returned
+        :type addons: list
+
+        :return: list of filtered elements
+        :rtype: list(KickstartElement)
+        """
+
+        filtered_elements = []
+        for element in self._elements:
+            if element.is_command():
+                if commands and element.name in commands:
+                    filtered_elements.append(element)
+            elif element.is_addon():
+                if addons and element.name in addons:
+                    filtered_elements.append(element)
+            elif element.is_section():
+                if sections and element.name in sections:
+                    filtered_elements.append(element)
+        return filtered_elements
+
+    @staticmethod
+    def get_kickstart_from_elements(elements=None):
+        """Returns kickstart generated from elements.
+
+        :param elements: list of kickstart elements for generated kickstart
+        :type elements: list(KickstartElement)
+
+        :return: kickstart
+        :rtype: str
+        """
+        return "".join(element.content for element in elements)
+
+    def __str__(self):
+        return str(self._elements)
+
+
+class TrackedKickstartElements(KickstartElements):
+    """Container for kickstart elements with tracking."""
+    def __init__(self):
+        super().__init__()
+        self._processed_elements = set()
+
+    def get_and_process_elements(self, commands=None, sections=None, addons=None):
+        """Returns selected elements and marks them as processed.
+
+        :param commands: names of commands to be returned
+        :type commands: list
+        :param sections: names of sections to be returned
+        :type sections: list
+        :param addons: names of addons to be returned
+        :type addons: list
+
+        :return: list of filtered elements
+        :rtype: list(KickstartElement)
+        """
+        elements = self.get_elements(commands, sections, addons)
+        self._processed_elements.update(elements)
+        return elements
+
+    @property
+    def unprocessed_elements(self):
+        """List of all elements not tracked as processed.
+
+        :return: list of all elements not tracked as processed
+        :rtype: list(KickstartElement)
+        """
+        return [element for element in self._elements
+                if element not in self._processed_elements]
```

Comment:
I would change this to:
```
elements = set(self._elements)
return self._processed_elements - elements
```
Or even better change the `self._elements` parent's variable to `set`.

Reply:
That would not preserve the order of elements (also we are appending, not adding elements to the container).

Reply:
Ohh ok, I didn't know the order is required here too.